### PR TITLE
Disable TempRValueOpt when there are users of the temp object preceding the copy

### PR DESCRIPTION
**Pull request by meg-gupta**
rdar://108361427

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Review by atrick — APPROVED** [review]
LGTM

I got a little hung up trying to figure out how you can assume that projections are in the copy block. So it might be nice to comment that we assume collectLoads will enforce that. It could be in a different PR.

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 1]
```diff
@@ -542,6 +542,8 @@ void TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
   // re-initialization of the source.
   bool needFinalDeinit = copyInst->isTakeOfSrc();
 
+  SmallPtrSet<SILInstruction *, 4> userSet;
```
Comment (nate-chandler) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp:545]: Nit: `InstructionSet`?

**Review by nate-chandler — APPROVED** [review]
lgtm

**Comment by meg-gupta** [comment]
@swift-ci smoke test and merge

**Comment by meg-gupta** [comment]
@swift-ci apple silicon benchmark

**Pull request merged**